ravenml/data/write_dataset.py

```diff
@@ -0,0 +1,248 @@
+            path (Path object): target directory
+        """
+        if os.path.isdir(path):
+            print('\n', "Deleting contents of", path,
+                "you've got five seconds to cancel this.")
+            time.sleep(5)
+
+            for the_file in os.listdir(path):
+                file_path = os.path.join(path, the_file)
+                try:
+                    if os.path.isfile(file_path):
+                        os.unlink(file_path)
+                    elif os.path.isdir(file_path):
+                        shutil.rmtree(file_path)
+                except Exception as e:
+                    print(e)
+                    raise Exception(
+                        "Directory deletion failed. Maybe you have a file explorer/terminal open in this dir?"
+                    )
+
+            print('\n', "Done deleting contents of", path)
+
+    def split_data(self, obj_list):
+        """Splits obj_list into test/dev sets
+        
+        Args:
+            obj_list (list): list of objects to divide into test/dev
+            test_percent (int): percentage of objects in the test set
+        
+        Returns:
+            tuple of two lists. The first list is test, second dev
+        """
+        if len(obj_list) == 0:
+            raise Exception("Empty object list passed.")
+        if len(obj_list) == 1:
+            raise Exception(
+                "Object list of length 1 passed. Can't build test and dev set with this."
+            )
+
+        shuffle(obj_list)
+        index_to_split_on = max(1, int(len(obj_list) * self.test_percent))
+
+        test = obj_list[:index_to_split_on]
+        dev = obj_list[index_to_split_on:]
+
+        return (test, dev)
+
+    def write_related_data(self, objects, path):
+        if not os.path.exists(path):
+            os.makedirs(path)
+
+        for obj in objects:
+            self.copy_associated_files(path, obj)
+
+    def copy_associated_files(self, destination, obj):
+            data_dir = self.cache.path / 'data'
+
+            for suffix in self.associated_files.values():
+                for prefix in self.related_data_prefixes.values():
+                    filepath = data_dir / f'{prefix}{obj.get("image_id")}{suffix}'
+                    if os.path.exists(filepath):
+                        shutil.copy(
+                            str(filepath.absolute()), str(destination.absolute()))
+
+    def write_out_complete_set(self, path, data):
+        """Writes out the special complete set into 'dev'. No validation data.
+        
+        Args:
+            path (Path): directory to write complete set out to
+            data (list): objects to write to the complete set
+        """
+        record_path = path / 'train'
+        if not os.path.exists(record_path):
+            os.makedirs(record_path)
+
+        test_record_data, train_record_data = self.split_data(data)
+
+        self.write_out_tf_examples(train_record_data, record_path / 'train.record')
```

Comment:
here again it is TF specific. I imagine there must be a way to make this functionality platform independent. We could break out TF specific functionality into a `tf.py` module or something, whose functions plugins can use to define library-independent versions of something like `write_out_complete_set` and `export_as_TFExample`.

I realize this is more complex and our current needs don't require it. But for this pipeline to meet its intended goals, its important that the dataset creation architecture is as generic as possible in the same way the training architecture is. We can also include more specific functionality that serves our purposes, but it should be built off something generic.

That's just my biased opinion because I want this pipeline to become the best open source ML architecture out there, and limiting support to just TF is not good for that